// @forge/storage/out/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIError = exports.SortOrder = exports.FilterConditions = exports.WhereConditions = exports.startsWith = exports.GlobalStorage = exports.getStorageInstanceWithQuery = void 0;
const entity_storage_1 = require("./entity-storage");
const query_api_1 = require("./query-api");
const getStorageInstanceWithQuery = (adapter) => {
    return {
        get: adapter.get.bind(adapter),
        set: adapter.set.bind(adapter),
        delete: adapter.delete.bind(adapter),
        getSecret: adapter.getSecret.bind(adapter),
        setSecret: adapter.setSecret.bind(adapter),
        deleteSecret: adapter.deleteSecret.bind(adapter),
        query: () => new query_api_1.DefaultQueryBuilder(adapter),
        entity: (entityName) => new entity_storage_1.EntityStorageBuilder(entityName, adapter)
    };
};
exports.getStorageInstanceWithQuery = getStorageInstanceWithQuery;
var global_storage_1 = require("./global-storage");
Object.defineProperty(exports, "GlobalStorage", { enumerable: true, get: function () { return global_storage_1.GlobalStorage; } });
var conditions_1 = require("./conditions");
Object.defineProperty(exports, "startsWith", { enumerable: true, get: function () { return conditions_1.startsWith; } });
var conditions_2 = require("./eap/conditions");
Object.defineProperty(exports, "WhereConditions", { enumerable: true, get: function () { return conditions_2.WhereConditions; } });
Object.defineProperty(exports, "FilterConditions", { enumerable: true, get: function () { return conditions_2.FilterConditions; } });
var query_interfaces_1 = require("./query-interfaces");
Object.defineProperty(exports, "SortOrder", { enumerable: true, get: function () { return query_interfaces_1.SortOrder; } });
var errors_1 = require("./errors");
Object.defineProperty(exports, "APIError", { enumerable: true, get: function () { return errors_1.APIError; } });
