// @forge/ui/out/components.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueContext = exports.IssueGlance = exports.IssueAction = exports.MacroConfig = exports.Macro = exports.RadioGroup = exports.Radio = exports.ThreeLOPrompt = exports.Select = exports.UserPicker = exports.Option = exports.PortalTextArea = exports.TextField = exports.Tooltip = exports.Toggle = exports.TextArea = exports.StatusLozenge = exports.SectionMessage = exports.Range = exports.ModalDialog = exports.FormCondition = exports.Form = exports.Table = exports.Row = exports.Head = exports.Cell = exports.Heading = exports.Tabs = exports.Tab = exports.Fragment = exports.ErrorPanel = exports.DatePicker = exports.DateLozenge = exports.CheckboxGroup = exports.Checkbox = exports.InlineDialog = exports.Image = exports.Link = exports.Strike = exports.Em = exports.Strong = exports.Text = exports.TagGroup = exports.Tag = exports.Code = exports.ButtonSet = exports.Button = exports.Badge = exports.UserGroup = exports.User = void 0;
exports.AssetsAppImportTypeConfiguration = exports.DashboardGadgetEdit = exports.DashboardGadget = exports.PortalUserMenuAction = exports.PortalRequestViewAction = exports.PortalProfilePanel = exports.PortalFooter = exports.PortalSubheader = exports.PortalHeader = exports.OrganizationPanel = exports.PortalRequestDetailPanel = exports.PortalRequestDetail = exports.QueuePage = exports.ProjectSettingsPage = exports.ProjectPage = exports.AdminPage = exports.CustomFieldContextConfig = exports.PortalRequestCreatePropertyPanel = exports.CustomFieldEdit = exports.CustomField = exports.IssueActivity = exports.IssuePanel = exports.IssuePanelAction = void 0;
const tslib_1 = require("tslib");
const _1 = tslib_1.__importStar(require("./"));
const types_1 = require("./types");
const portaRequestCreatePropertyPanelTypes_1 = require("./portaRequestCreatePropertyPanelTypes");
const useProductContext_1 = require("./hooks/useProductContext");
exports.User = 'User';
exports.UserGroup = 'UserGroup';
exports.Badge = 'Badge';
exports.Button = 'Button';
exports.ButtonSet = 'ButtonSet';
exports.Code = 'Code';
exports.Tag = 'Tag';
exports.TagGroup = 'TagGroup';
exports.Text = 'Text';
exports.Strong = 'Strong';
exports.Em = 'Em';
exports.Strike = 'Strike';
exports.Link = 'Link';
exports.Image = 'Image';
exports.InlineDialog = 'InlineDialog';
exports.Checkbox = 'Checkbox';
exports.CheckboxGroup = 'CheckboxGroup';
exports.DateLozenge = 'DateLozenge';
exports.DatePicker = 'DatePicker';
exports.ErrorPanel = 'ErrorPanel';
exports.Fragment = 'Fragment';
exports.Tab = 'Tab';
exports.Tabs = 'Tabs';
exports.Heading = 'Heading';
exports.Cell = 'Cell';
exports.Head = 'Head';
exports.Row = 'Row';
exports.Table = 'Table';
exports.Form = 'Form';
exports.FormCondition = 'FormCondition';
exports.ModalDialog = 'ModalDialog';
exports.Range = 'Range';
exports.SectionMessage = 'SectionMessage';
exports.StatusLozenge = 'StatusLozenge';
exports.TextArea = 'TextArea';
exports.Toggle = 'Toggle';
exports.Tooltip = 'Tooltip';
exports.TextField = 'TextField';
exports.PortalTextArea = 'PortalTextArea';
exports.Option = 'Option';
exports.UserPicker = 'UserPicker';
exports.Select = 'Select';
exports.ThreeLOPrompt = 'ThreeLOPrompt';
exports.Radio = 'Radio';
exports.RadioGroup = 'RadioGroup';
const Macro = (props) => {
    return _1.default.createElement(exports.Fragment, null, props.app);
};
exports.Macro = Macro;
exports.MacroConfig = 'MacroConfig';
exports.IssueAction = 'IssueAction';
exports.IssueGlance = 'IssueGlance';
exports.IssueContext = 'IssueContext';
exports.IssuePanelAction = 'IssuePanelAction';
const IssuePanelPrimitive = 'IssuePanel';
const IssuePanel = ({ actions = [], children = [] }) => (_1.default.createElement(IssuePanelPrimitive, null, [...actions, ...(Array.isArray(children) ? children : [children])]));
exports.IssuePanel = IssuePanel;
exports.IssueActivity = 'IssueActivity';
const CustomFieldPrimitive = 'CustomField';
const CustomField = (props) => {
    return _1.default.createElement(CustomFieldPrimitive, null, props.children);
};
exports.CustomField = CustomField;
const CustomFieldEditPrimitive = 'CustomFieldEdit';
const PortalRequestCreatePropertyPanelPrimitive = 'PortalRequestCreatePropertyPanel';
const CustomFieldEdit = ({ children, onSubmit, width = 'medium', header = 'Custom field edit' }) => {
    const extensionContext = (0, useProductContext_1.useProductContext)().extensionContext;
    const incomingFieldValue = extensionContext && (0, types_1.isCustomFieldExtensionContext)(extensionContext) && extensionContext.fieldValue;
    const isInline = extensionContext && (0, types_1.isCustomFieldExtensionContext)(extensionContext) && extensionContext.isInline;
    const initialValue = incomingFieldValue != null && typeof incomingFieldValue !== 'boolean' ? incomingFieldValue : null;
    const [fieldValue, setFieldValue] = (0, _1.useState)({
        value: initialValue,
        updated: false
    });
    const [isOpen, setIsOpen] = (0, _1.useState)(true);
    const closeModal = () => setIsOpen(false);
    const onSaveFn = async (formValue) => {
        const value = await onSubmit(formValue);
        setFieldValue({
            value,
            updated: true
        });
    };
    return (_1.default.createElement(CustomFieldEditPrimitive, { fieldValue: fieldValue, isModalOpen: isOpen }, isInline ? (_1.default.createElement(exports.Form, { onSubmit: onSaveFn, submitButtonText: "Save" }, children)) : (_1.default.createElement(exports.ModalDialog, { header: header, onClose: closeModal, width: width },
        _1.default.createElement(exports.Form, { onSubmit: onSaveFn, submitButtonText: "Save" }, children)))));
};
exports.CustomFieldEdit = CustomFieldEdit;
const PortalRequestCreatePropertyPanel = ({ children, data }) => {
    const [fieldValue, setFieldValue] = (0, _1.useState)({
        value: { data: null, isValid: false },
        updated: false
    });
    const validateCustomPortalFieldValue = (formdata) => {
        const isValid = (formdata === null || formdata === void 0 ? void 0 : formdata.data) !== undefined;
        return isValid;
    };
    const ErrMessages = (errMsg) => {
        throw Error(errMsg);
    };
    (0, _1.useEffect)(() => {
        if (typeof data === 'object') {
            if (Object.keys(data).length != 0) {
                validateCustomPortalFieldValue(data)
                    ? setFieldValue({ value: data, updated: true })
                    : ErrMessages(portaRequestCreatePropertyPanelTypes_1.MISSING_FIELDS);
            }
        }
        else {
            ErrMessages(portaRequestCreatePropertyPanelTypes_1.INVALID_APP_DATA);
        }
    }, [data]);
    return (_1.default.createElement(PortalRequestCreatePropertyPanelPrimitive, { fieldValue: fieldValue }, children));
};
exports.PortalRequestCreatePropertyPanel = PortalRequestCreatePropertyPanel;
const CustomFieldContextConfigPrimitive = 'CustomFieldContextConfig';
const CustomFieldContextConfig = ({ children, onSubmit }) => {
    const [data, onFormSubmit] = (0, _1.useAction)(async (_, formData) => onSubmit(formData), undefined);
    const noop = () => undefined;
    const actionButtons = [_1.default.createElement(exports.Button, { text: "Cancel", appearance: "link", onClick: noop })];
    return (_1.default.createElement(CustomFieldContextConfigPrimitive, { data: data },
        _1.default.createElement(exports.Form, { submitButtonAppearance: "primary", submitButtonText: "Save", actionButtons: actionButtons, onSubmit: onFormSubmit }, children)));
};
exports.CustomFieldContextConfig = CustomFieldContextConfig;
exports.AdminPage = 'AdminPage';
exports.ProjectPage = 'ProjectPage';
exports.ProjectSettingsPage = 'ProjectSettingsPage';
exports.QueuePage = 'QueuePage';
exports.PortalRequestDetail = 'PortalRequestDetail';
exports.PortalRequestDetailPanel = 'PortalRequestDetailPanel';
exports.OrganizationPanel = 'OrganizationPanel';
exports.PortalHeader = 'PortalHeader';
exports.PortalSubheader = 'PortalSubheader';
exports.PortalFooter = 'PortalFooter';
exports.PortalProfilePanel = 'PortalProfilePanel';
exports.PortalRequestViewAction = 'PortalRequestViewAction';
exports.PortalUserMenuAction = 'PortalUserMenuAction';
exports.DashboardGadget = 'DashboardGadget';
const DashboardGadgetEditPrimitive = 'DashboardGadgetEdit';
const DashboardGadgetEdit = ({ children, onSubmit }) => {
    const { extensionContext } = (0, useProductContext_1.useProductContext)();
    const storedGadgetConfiguration = extensionContext && (0, types_1.isDashboardGadgetExtensionContext)(extensionContext) && extensionContext.gadgetConfiguration;
    const [formData, setFormData] = (0, _1.useState)(storedGadgetConfiguration || {});
    const onSaveFn = async (data) => {
        const dataToSave = await onSubmit(data);
        setFormData(dataToSave);
    };
    const noop = () => undefined;
    const actionButtons = [_1.default.createElement(exports.Button, { text: "Cancel", appearance: "link", onClick: noop })];
    return (_1.default.createElement(DashboardGadgetEditPrimitive, { formData: formData },
        _1.default.createElement(exports.Form, { onSubmit: onSaveFn, submitButtonText: "Save", actionButtons: actionButtons }, children)));
};
exports.DashboardGadgetEdit = DashboardGadgetEdit;
const AssetsAppImportTypeConfigurationPrimitive = 'AssetsAppImportTypeConfiguration';
const AssetsAppImportTypeConfiguration = ({ children, onSubmit }) => {
    const [saved, setSaved] = (0, _1.useState)(false);
    const onSaveFn = async (formValue) => {
        const result = await onSubmit(formValue);
        setSaved(result);
    };
    const noop = () => undefined;
    const actionButtons = [_1.default.createElement(exports.Button, { text: "Cancel", appearance: "subtle", onClick: noop })];
    return (_1.default.createElement(AssetsAppImportTypeConfigurationPrimitive, { saved: saved },
        _1.default.createElement(exports.Form, { onSubmit: onSaveFn, submitButtonText: "Save configuration", submitButtonAppearance: "primary", actionButtons: actionButtons }, children)));
};
exports.AssetsAppImportTypeConfiguration = AssetsAppImportTypeConfiguration;
